File: backend/GraficaModerna.Infrastructure/Services/CartService.cs
using GraficaModerna.Application.DTOs;
using GraficaModerna.Application.Interfaces;
using GraficaModerna.Domain.Entities;
using GraficaModerna.Infrastructure.Context;
using Microsoft.EntityFrameworkCore;

namespace GraficaModerna.Infrastructure.Services;

public class CartService : ICartService
{
    private readonly AppDbContext _context;

    public CartService(AppDbContext context)
    {
        _context = context;
    }

    private async Task<Cart> GetCartEntity(string userId)
    {
        var cart = await _context.Carts
            .Include(c => c.Items).ThenInclude(i => i.Product)
            .FirstOrDefaultAsync(c => c.UserId == userId);

        if (cart == null)
        {
            cart = new Cart { UserId = userId };
            _context.Carts.Add(cart);
            await _context.SaveChangesAsync();
        }
        return cart;
    }

    public async Task<CartDto> GetCartAsync(string userId)
    {
        var cart = await GetCartEntity(userId);
        var itemsDto = cart.Items.Select(i => new CartItemDto(
            i.Id,
            i.ProductId,
            i.Product?.Name ?? "Indisponível",
            i.Product?.ImageUrl ?? "",
            i.Product?.Price ?? 0,
            i.Quantity,
            (i.Product?.Price ?? 0) * i.Quantity
        )).ToList();

        return new CartDto(cart.Id, itemsDto, itemsDto.Sum(i => i.TotalPrice));
    }

    public async Task AddItemAsync(string userId, AddToCartDto dto)
    {
        var product = await _context.Products.FindAsync(dto.ProductId);
        if (product == null) throw new Exception("Produto não encontrado.");
        if (!product.IsActive) throw new Exception("Produto indisponível.");
        if (product.StockQuantity < dto.Quantity) throw new Exception($"Estoque insuficiente. Restam {product.StockQuantity}.");

        var cart = await GetCartEntity(userId);
        var existing = cart.Items.FirstOrDefault(i => i.ProductId == dto.ProductId);

        if (existing != null)
        {
            if (product.StockQuantity < (existing.Quantity + dto.Quantity))
                throw new Exception("Estoque insuficiente para adicionar mais.");
            existing.Quantity += dto.Quantity;
        }
        else
        {
            cart.Items.Add(new CartItem { CartId = cart.Id, ProductId = dto.ProductId, Quantity = dto.Quantity });
        }
        cart.LastUpdated = DateTime.UtcNow;
        await _context.SaveChangesAsync();
    }

    public async Task RemoveItemAsync(string userId, Guid itemId)
    {
        var cart = await GetCartEntity(userId);
        var item = cart.Items.FirstOrDefault(i => i.Id == itemId);
        if (item != null)
        {
            _context.CartItems.Remove(item);
            await _context.SaveChangesAsync();
        }
    }

    public async Task ClearCartAsync(string userId)
    {
        var cart = await GetCartEntity(userId);
        _context.CartItems.RemoveRange(cart.Items);
        await _context.SaveChangesAsync();
    }
}

File: grafica-moderna/app/Services/CartService.php
<?php

namespace App\Services;

use App\Models\Cart;
use App\Models\CartItem;
use App\Models\Product;
use Illuminate\Support\Facades\DB;
use Illuminate\Support\Facades\Log;
use Exception;

class CartService
{
    const MAX_QUANTITY_PER_ITEM = 5000;

    public function getCart(string $userId)
    {
        // GetOrCreate Logic
        $cart = Cart::firstOrCreate(
            ['user_id' => $userId],
            ['last_updated' => now()]
        );

        // Carrega items com produtos
        $cart->load('items.product');

        // Limpeza de órfãos (produtos deletados ou inativos)
        $orphans = $cart->items->filter(fn($item) => !$item->product || !$item->product->is_active);
        
        if ($orphans->isNotEmpty()) {
            CartItem::destroy($orphans->pluck('id'));
            $cart->refresh(); // Recarrega após deletar
        }

        // Formatação do DTO de resposta
        $itemsDto = $cart->items->map(function ($item) {
            $product = $item->product;
            // Assumindo que image_urls é array no cast do model Product
            $mainImage = !empty($product->image_urls) ? $product->image_urls[0] : "";

            return [
                'id' => $item->id,
                'productId' => $item->product_id,
                'productName' => $product->name,
                'imageUrl' => $mainImage,
                'unitPrice' => $product->price,
                'quantity' => $item->quantity,
                'totalPrice' => $product->price * $item->quantity,
                'weight' => $product->weight,
                'dimensions' => [
                    'width' => $product->width,
                    'height' => $product->height,
                    'length' => $product->length
                ]
            ];
        });

        return [
            'id' => $cart->id,
            'items' => $itemsDto,
            'totalAmount' => $itemsDto->sum('totalPrice')
        ];
    }

    public function addItem(string $userId, array $dto)
    {
        if ($dto['quantity'] <= 0) throw new Exception("Quantidade inválida.");
        if ($dto['quantity'] > self::MAX_QUANTITY_PER_ITEM) throw new Exception("Limite de quantidade excedido.");

        // Laravel gerencia transações e retries de deadlock automaticamente (3 tentativas padrão)
        DB::transaction(function () use ($userId, $dto) {
            // LockForUpdate para evitar condição de corrida no estoque
            $product = Product::where('id', $dto['productId'])->lockForUpdate()->first();

            if (!$product) throw new Exception("Produto indisponível.");
            if ($product->stock_quantity < $dto['quantity']) throw new Exception("Estoque insuficiente.");

            $cart = Cart::firstOrCreate(['user_id' => $userId]);

            $existingItem = CartItem::where('cart_id', $cart->id)
                ->where('product_id', $dto['productId'])
                ->first();

            if ($existingItem) {
                $newTotal = $existingItem->quantity + $dto['quantity'];
                
                if ($newTotal > self::MAX_QUANTITY_PER_ITEM) throw new Exception("Limite excedido.");
                if ($product->stock_quantity < $newTotal) throw new Exception("Estoque insuficiente para o total.");

                $existingItem->quantity = $newTotal;
                $existingItem->save();
            } else {
                CartItem::create([
                    'cart_id' => $cart->id,
                    'product_id' => $dto['productId'],
                    'quantity' => $dto['quantity']
                ]);
            }

            $cart->update(['last_updated' => now()]);
        }, 3); // 3 tentativas
    }

    public function updateItemQuantity(string $userId, string $itemId, int quantity)
    {
        if (quantity < 0) throw new Exception("Qtd negativa.");
        if (quantity == 0) {
            $this->removeItem($userId, $itemId);
            return;
        }

        DB::transaction(function () use ($userId, $itemId, $quantity) {
            $cart = Cart::where('user_id', $userId)->firstOrFail();
            
            $item = CartItem::where('id', $itemId)
                ->where('cart_id', $cart->id)
                ->firstOrFail();

            $product = Product::where('id', $item->product_id)->lockForUpdate()->first();

            if ($product->stock_quantity < $quantity) {
                throw new Exception("Estoque insuficiente.");
            }

            $item->quantity = $quantity;
            $item->save();
            $cart->touch(); // Atualiza updated_at/last_updated
        });
    }

    public function removeItem(string $userId, string $itemId)
    {
        $cart = Cart::where('user_id', $userId)->first();
        if ($cart) {
            CartItem::where('cart_id', $cart->id)->where('id', $itemId)->delete();
            $cart->touch();
        }
    }

    public function clearCart(string $userId)
    {
        $cart = Cart::where('user_id', $userId)->first();
        if ($cart) {
            $cart->items()->delete();
            $cart->touch();
        }
    }
}